File: 1458-sort-integers-by-the-number-of-1-bits/sort-integers-by-the-number-of-1-bits.js
/**
 * @param {number[]} arr
 * @return {number[]}
 */
var sortByBits = function(arr) {
    const buckets = new Array(32);

    for(let i=0; i<buckets.length; i++) {
        buckets[i] = [];
    }

    function countBits(val) {
        let count = 0;
        for(let i=0; i<=31; i++) {
            if((val & (1 << i)) != 0) count++;
        }
        return count;
    }

    for(let i=0; i<arr.length; i++) {
        let c = countBits(arr[i]);
        buckets[c].push(arr[i]);
    }

    for(let i=0; i<buckets.length; i++) {
        buckets[i].sort((a, b) => a - b);
    }

    const ans = [];

    for(let i=0; i<buckets.length; i++) {
        for(val of buckets[i]) ans.push(val);
    }
    return ans;
};

File: 1458-sort-integers-by-the-number-of-1-bits/sort-integers-by-the-number-of-1-bits.java
class Solution {
    public int[] sortByBits(int[] arr) {
        List<Integer>[] buckets = new ArrayList[32];
        for(int i=0; i<32; i++) {
            buckets[i] = new ArrayList<Integer>();
        }
        for(int val : arr) {
            buckets[Integer.bitCount(val)].add(val);
        }
        for(int i=0; i<32; i++) {
            Collections.sort(buckets[i]);
        }
        int i=0;
        for(int j=0; j<32; j++) {
            for(int val : buckets[j]) {
                arr[i++] = val;
            }
        }
        return arr;
    }
}